docs(README): Expand details for devs/contributors + misc tweaks (#541)
nextcloud/updater: README.md

* [*What does the updater do?*](https://docs.nextcloud.com/server/latest/admin_manual/maintenance/update.html#what-does-the-updater-do)

> [!NOTE]
> The built-in Updater is primarily applicable to manual/Archive (aka: "tarball / zip") installations. Most other installation methods (such as Docker images and Snaps) utilize their own officially supported processes for keeping Nextcloud Server up-to-date.
> **Please follow their respective documented approaches rather than trying to run the Updater yourself.**

## Installation

### Requirements

* An Archive based installation (i.e. **not** Docker, Snap, AIO, etc. unless you're the packager of said method and utilize the Updater internally or permit your users to utilize it).
* The app is bundled with the installation Archive for Nextcloud Server. No additional steps are necessary to install.

Note:

* The Updater "app" is not a standard Nextcloud Server app so it is **not** managed via the app store.

> * The Web UI mode of Updater will not be active unless the `updatenotification` app is enabled.

Reviewer: Out of curiosity, is there a reason for this? Is the web UI actually part of the other app?

Reviewer: Yes, this is answered in the rest of the README.

The following are standard parameters already set in any functioning Server installation, but which the Updater requires:

* [`datadirectory`]
* [`instanceid`]

Other update related parameters:

* [`logfile`](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#logfile)
* [`updatechecker`](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#updatechecker)
* [`appcodechecker`](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#appcodechecker)

## Usage

Using the Updater via the Web UI is more convenient, but using it from the command line is more reliable.

> [!WARNING]
> Please make a [backup](https://docs.nextcloud.com/server/latest/admin_manual/maintenance/backup.html) of your data and database and familiarize yourself with the [restore](https://docs.nextcloud.com/server/latest/admin_manual/maintenance/restore.html) process before proceeding with using the Updater.

### Web UI mode

> Go to *Administration settings*. It's can be reached in the *Overview* under the *Update* heading.

Reviewer: ```suggestion
Go to *Administration settings*. It can be reached in the *Overview* under the *Update* heading.
```

The Updater app handles step #1 (see [*What does the updater do?*](https://docs.nextcloud.com/server/latest/admin_manual/maintenance/update.html#what-does-the-updater-do)) and prepares the environment for steps #2 and #3 (which are handled by `occ upgrade` *after* the Updater successfully completes).

The Updater app needs to function while the Nextcloud Server is offline for code updates. It is therefore designed to operate as independently as possible from the rest of Server. This permits Server to be updated safely and the Updater to proceed without interruption during code swaps. The trade-off is that none of the typical Nextcloud classes/APIs are available to serve the needs of the Updater. The Updater, fortunately, has a relatively narrowly defined mission and set of needs.

### Theory of Operation

* The Updater runs from either `/updater/index.php` (Web mode) or `/updater/updater.phar` (Command line mode), which are
  provided by the previous (last/current) Server installation.
* The Updater maintains state in `.step` files/folders
* Connecting clients are provided with temporary failure HTTP codes to minimize end user interruptions

> Some historical more in-depeth context can be found in nextcloud/updater#1 and nextcloud/updater#2.

Reviewer: ```suggestion
Some historical more in-depth context can be found in nextcloud/updater#1 and nextcloud/updater#2.
```

--no-upgrade             Don't automatically run `occ upgrade` when the Updater finishes (note: `occ upgrade` is required after Updater updates the application code in order to push out any database changes in the newly deployed version of Nextcloud)
```

## How it works

Keeping Nextcloud Server up-to-date, at a high level, consists of three things:

1. Updating the code
2. Upgrading the database
3. Updating all shipped and locally deployed apps

The Updater app handles step #1 (see [*What does the updater do?*](https://docs.nextcloud.com/server/latest/admin_manual/maintenance/update.html#what-does-the-updater-do)) and prepares the environment for steps #2 and #3 (which are handled by `occ upgrade` *after* the Updater successfully completes).

> The Updater app is designed to be as independent from Server as possible for reliability purposes.

Reviewer: Not for reliability, but because it needs to run while the server is down because its code is being updated.

Author: Agreed. That's what I meant. :smile:  But you're right: I chose ambiguous wording. I've revised.

### Theory of Operation

> * The Updater is wholly independent of the Server installation (save for being ran from `/updater/index.php` or `/updater/updater.phar`, which are

Reviewer: ```suggestion
* The Updater is wholly independent of the Server installation (safe for being ran from `/updater/index.php` or `/updater/updater.phar`, which are
```
This line is clearer, so I would remove the one above about being «as independent from Server as possible»

Author: ```suggestion
* The Updater runs from either `/updater/index.php` (Web mode) or `/updater/updater.phar` (Command line mode), which are
```

* Connecting clients are provided with temporary failure HTTP codes to minimize end user interruptions

Some historical more in-depeth context can be found in nextcloud/updater#1 and nextcloud/updater#2.

### Components

#### Updater app components

- `/updater/updater.phar` (CLI Updater)
- `/updater/index.php` (Web Updater)

#### Server components

- `occ upgrade` (Database Upgrader)
- Web `occ upgrade` trigger/monitor
- `updatenotification` app which provides the *Update* overvew under *Administration settings* and handles notifying admins of the availability of updates
- The legacy update availability notifier (only activated if the `updatenotification` is disabled and does not provide an Overview screen; currently generates an indefinite "nag")

### Update methods

Since there are multiple ways of triggering and using the Updater as well as the later database upgrades and app updates, it is useful to have an overview of the different work flows. This is particularly useful if you're doing development related to the Nextcloud Server update/upgrade process. It also further shows how steps 1-3 above are coupled together.

#### Semi-Manual

This approach 

Consists of:

> * Manually replacing installation folder code contents except for config, data, and locally installed apps

Author: ```suggestion
* Manually replacing installation folder code contents except for config, data, and locally installed apps (i.e. does **not** use the Updater app at all)
```

		- https://github.com/nextcloud/server/blob/master/lib/private/Updater.php
	- Handles database upgrade migrations
	- Handles app updates (i.e. for compatibility with a new major version of Server)

### Dependences needed for building

box, composer, etc. (to be filled in later)

### Build artifacts / What to check in

#### For Distribution

- `/updater.phar`
- `/index.php`

#### For Check-in (same + implementation changes)

- `/updater.phar`
- `/index.php`

Plus whatever has been changed in the implementation in:

- `/lib/*`
- `/index.web.php`
- `/updater.php`

#### Transient

Used during the build process but not checked in:

* Specific to the CLI Updater:
  - `/lib/Version.php`

### Testing

#### CI

(to be filled in)

#### Unit/other

(to be filled in)

#### Locally/manually

(to be filled in)

## Troubleshooting

### Logging

* The Updater operates independently from Server so it has it's own log file called `updater.log` which is located in the configured `datadirectory` for the instance.
* The database upgrade migrations and app updates are not handled directly by the Updater so they're logged in the standard Nextcloud log (`nextcloud.log`)

> ### Web-based Updater isn't completely successfully or reporting on what is going on

Reviewer: ```suggestion
### Web-based Updater isn't completely successfull or reporting on what is going on
```
